### ISSUE TITLE
Focus issue when using inputs within an overlay

### ISSUE THREAD

**Issue by Benji1** [adobe/react-spectrum#2380]
# 🐛 Bug Report 

Hey react-spectrum team, thanks for your awesome work! I noticed that when clicking on a switch within an overlay, the useFocusRing hook returns isFocusVisible true even if it should not, which then leads to a visible focus ring. I noticed something similar here: https://github.com/adobe/react-spectrum/issues/1050 and a workaround is provided there. However, after using the workaround and adding a textfield to the overlay, it will not focus anymore when clicking on it.
I would be really great if the hook would also return the correct value in these circumstances.
A fix for this would be highly appreciated.

## 🤔 Expected Behavior

isFocusVisible should not return true in that case, since it is a mouse click and not a keyboard event triggering it.

## 😯 Current Behavior

isFocusVisible returns a wrong value in the above described cases.

## 🔦 Context

I want to use a switch inside an overlay without it getting focused when clicking on it and without preventing the textfields from getting focused.

## 💻 Code Sample

Code mostly taken from examples of https://react-spectrum.adobe.com/react-aria, reproducible case here:
https://codesandbox.io/s/react-aria-mouse-focus-issue-38tqd

![focus](https://user-images.githubusercontent.com/7806108/134930552-077c1ab0-5f3a-4afd-b26b-feb8d13952e0.png)
## 🌍 Your Environment

Codesandbox

| Software         | Version(s) |
| ---------------- | ---------- |
| react-spectrum   | see below
| Browser          | latest FF & Chrome
| Operating System | Linux/Win

    "react": "17.0.2" 
    "react-dom": "17.0.2",
    "react-scripts": "4.0.0",
    "@react-aria/overlays": "^3.7.2",
    "@react-aria/dialog": "^3.1.4",
    "@react-aria/focus": "^3.5.0",
    "@react-aria/button": "^3.3.4",
    "@react-aria/utils": "^3.9.0",
    "@react-aria/switch": "^3.1.3",
    "@react-aria/textfield": "^3.4.0"

**Comment by nimrod13** [comment]
The same thing happens for buttons inside an overlay. So relying on isFocusVisible or isFocused in these cases would not work.

**Comment by reidbarber** [comment]
Looks like it the issue might be within `usePress` then.

**Comment by nimrod13** [comment]
Maybe it's something related to `isVirtualClick()`? I also noticed that `isKeyboardFocusEvent()` returns true when clicking on the switch (probably due to the undefined textInput prop), so maybe there should also be an additional check: `isTextInput !== undefined `

**Comment by reidbarber** [comment]
Yeah, it looks like @LFDanLu might have caught this back in https://github.com/adobe/react-spectrum/issues/1050#issuecomment-687315637. I'll take a closer look and see if I can come up with a potential solution.

**Comment by Benji1** [comment]
Thanks for your fast answers!

**Comment by LFDanLu** [comment]
Did some more digging, and I think there are couple things that will need to be investigated/considered here:
1. How to not set currentModality to "virtual" when the user clicks on the label of an input. Not entirely sure if we can differentiate that case from virtual clicks from accessibility tools or element.click() but it would help with this specific bug since the aria example hides the input via VisuallyHidden so the user is actually clicking on the label of the switch.  

2. Having useTextField handle focusing itself on press. Other RSP components have a usePress inside of it that handles this but TextField doesn't want usePress necessarily since it will prevent default behavior such as text selection via mouse. The workaround mentioned in https://github.com/adobe/react-spectrum/issues/1050#issuecomment-687315637 breaks focusing the textfield in the popover via click, but would no longer be a problem if we had useTextField handle focusing the input on pointer down via `onPointerDown={() => ref.current.focus()}`

**Comment by ritz078** [comment]
I am facing the same issue with `useRadio`.

**Comment by ritz078** [comment]
Facing the same issue with `useRadio`. Is this a bug that is still considered ?
